File: backend/app/schemas/portal_config.py
from pydantic import BaseModel, Field, field_validator


class SpaceConfig(BaseModel):
    id: int
    name: str
    file_count: int = 0
    tag_count: int = 0
    enabled: bool = True


class DomainConfig(BaseModel):
    name: str
    space_ids: list[int] = Field(default_factory=list)
    color: str
    bg: str
    icon: str
    background_image: str = ""
    enabled: bool = True

    @field_validator("space_ids")
    @classmethod
    def validate_space_ids(cls, value: list[int]) -> list[int]:
        if not value:
            raise ValueError("space_ids cannot be empty")
        return value


class SectionConfig(BaseModel):
    title: str
    tag: str
    link: str
    icon: str
    color: str = "#2563eb"
    bg: str = "#eff6ff"
    enabled: bool = True


class QAConfig(BaseModel):
    knowledge_space_ids: list[int] = Field(default_factory=list)
    panel_title: str = "技术问答·专家在线"
    welcome_message: str = "你好，我是首钢知库智能助手，请问有什么可以帮您？"
    hot_questions: list[str] = Field(default_factory=list)
    ai_search_system_prompt: str = ""
    qa_system_prompt: str = ""
    selected_model: str = ""


class QAModelOption(BaseModel):
    key: str = ""
    id: str
    name: str = ""
    display_name: str = ""
    visual: bool = False


class QAModelOptionsResponse(BaseModel):
    selected_model: str = ""
    models: list[QAModelOption] = Field(default_factory=list)


class SpaceOption(BaseModel):
    id: int
    name: str
    description: str = ""
    file_count: int = 0


class SpaceOptionsResponse(BaseModel):
    options: list[SpaceOption] = Field(default_factory=list)


class SpaceFileItem(BaseModel):
    id: int
    name: str


class SpaceFilesResponse(BaseModel):
    space_id: int
    files: list[SpaceFileItem] = Field(default_factory=list)


class RecommendationConfig(BaseModel):
    provider: str
    home_strategy: str
    detail_strategy: str


class DisplayHomeConfig(BaseModel):
    section_page_size: int = 6
    hot_tags_count: int = 8
    qa_hot_count: int = 4
    domain_count: int = 6
    spaces_count: int = 6
    apps_count: int = 6


class DisplayListConfig(BaseModel):
    page_size: int = 10
    visible_tag_count: int = 2


class DisplaySearchConfig(BaseModel):
    page_size: int = 10
    visible_tag_count: int = 2


class DisplayDetailConfig(BaseModel):
    related_files_count: int = 3
    visible_tag_count: int = 2


class DisplayConfig(BaseModel):
    home: DisplayHomeConfig
    list: DisplayListConfig
    search: DisplaySearchConfig
    detail: DisplayDetailConfig


class AppConfig(BaseModel):
    id: int
    name: str
    icon: str
    desc: str
    color: str
    bg: str
    url: str = ""
    enabled: bool = True


class PortalConfig(BaseModel):
    spaces: list[SpaceConfig] = Field(default_factory=list)
    domains: list[DomainConfig] = Field(default_factory=list)
    sections: list[SectionConfig] = Field(default_factory=list)
    qa: QAConfig
    recommendation: RecommendationConfig
    display: DisplayConfig
    apps: list[AppConfig] = Field(default_factory=list)


class SpacesConfigUpdate(BaseModel):
    spaces: list[SpaceConfig]


class DomainsConfigUpdate(BaseModel):
    domains: list[DomainConfig]


class SectionsConfigUpdate(BaseModel):
    sections: list[SectionConfig]


class AppsConfigUpdate(BaseModel):
    apps: list[AppConfig]

File: frontend/src/utils/portalConfig.ts
import type { AppConfig, BannerSlide, DisplayConfig, DomainConfig, SectionConfig, SpaceConfig } from '../api/adminConfig';
import { DISPLAY_CONFIG } from '../config/display';

export interface RuntimeBanner {
  label: string;
  title: string;
  desc: string;
  imageUrl: string;
  linkUrl: string;
}

export const FALLBACK_HOME_BANNERS: RuntimeBanner[] = [
  {
    label: '平台概览',
    title: '首钢知库 — 钢铁行业知识共享平台',
    desc: '汇聚设备、轧线、冷轧、能源全域知识，助力技术传承与创新',
    imageUrl: '/banner-hero-1.jpg',
    linkUrl: '',
  },
  {
    label: '专题推荐',
    title: '典型案例·事故分析专题上线',
    desc: '从实践中学习，从案例中成长，构建安全生产知识体系',
    imageUrl: '/banner-hero-2.jpg',
    linkUrl: '',
  },
  {
    label: '能力升级',
    title: '技术问答全新升级',
    desc: 'AI 驱动的智能问答系统，快速定位知识、精准解答技术难题',
    imageUrl: '/banner-hero-3.jpg',
    linkUrl: '',
  },
];

export function resolveHomeBanners(banners?: BannerSlide[]): RuntimeBanner[] {
  const portalBanners = (banners ?? [])
    .filter((banner) => banner.enabled && banner.image_url)
    .map<RuntimeBanner>((banner) => ({
      label: banner.label ?? '',
      title: banner.title ?? '',
      desc: banner.desc ?? '',
      imageUrl: banner.image_url,
      linkUrl: banner.link_url ?? '',
    }));
  return portalBanners.length ? portalBanners : FALLBACK_HOME_BANNERS;
}

export interface RuntimeDisplayConfig {
  home: {
    sectionPageSize: number;
    hotTagsCount: number;
    qaHotCount: number;
    domainCount: number;
    spacesCount: number;
    appsCount: number;
  };
  list: {
    pageSize: number;
    visibleTagCount: number;
  };
  search: {
    pageSize: number;
    visibleTagCount: number;
  };
  detail: {
    relatedFilesCount: number;
    visibleTagCount: number;
  };
}

export function toRuntimeDisplayConfig(display?: DisplayConfig): RuntimeDisplayConfig {
  if (!display) return DISPLAY_CONFIG;
  return {
    home: {
      sectionPageSize: display.home.section_page_size,
      hotTagsCount: display.home.hot_tags_count,
      qaHotCount: display.home.qa_hot_count,
      domainCount: display.home.domain_count,
      spacesCount: display.home.spaces_count,
      appsCount: display.home.apps_count,
    },
    list: {
      pageSize: display.list.page_size,
      visibleTagCount: display.list.visible_tag_count,
    },
    search: {
      pageSize: display.search.page_size,
      visibleTagCount: display.search.visible_tag_count,
    },
    detail: {
      relatedFilesCount: display.detail.related_files_count,
      visibleTagCount: display.detail.visible_tag_count,
    },
  };
}

export function getEnabledSpaces(spaces: SpaceConfig[]): SpaceConfig[] {
  return spaces.filter((space) => space.enabled);
}

export function getEnabledDomains(domains: DomainConfig[], spaces?: SpaceConfig[]): DomainConfig[] {
  const enabledSpaceIds = spaces ? new Set(getEnabledSpaces(spaces).map((space) => space.id)) : null;
  return domains.filter((domain) => {
    if (!domain.enabled || !domain.space_ids.length) return false;
    if (!enabledSpaceIds) return true;
    return domain.space_ids.some((spaceId) => enabledSpaceIds.has(spaceId));
  });
}

export function getEnabledSections(sections: SectionConfig[]): SectionConfig[] {
  return sections.filter((section) => section.enabled);
}

export function getEnabledApps(apps: AppConfig[]): AppConfig[] {
  return apps.filter((app) => app.enabled);
}

export function getPrimarySpaceId(spaceIds: number[]): number | undefined {
  return spaceIds.find((spaceId) => Number.isFinite(spaceId));
}
